core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1533,16 +1534,22 @@ object Parser {
   /** An empty iterable is the same as fail
     */
   def charIn(cs: Iterable[Char]): Parser[Char] =
-    if (cs.isEmpty) fail
-    else {
-      val ary = cs.toArray
-      Arrays.sort(ary)
-      rangesFor(ary) match {
-        case NonEmptyList((low, high), Nil) if low == Char.MinValue && high == Char.MaxValue =>
-          anyChar
-        case notAnyChar =>
-          Impl.CharIn(ary(0).toInt, BitSetUtil.bitSetFor(ary), notAnyChar)
-      }
+    cs match {
+      case _ if cs.isEmpty => fail
+      case range: NumericRange.Inclusive[Char] if range.step.toInt == 1 =>
+        val start = range.start.toInt
+        val end = range.end.toInt
+        val bitSet = BitSetUtil.bitSetForRange(end - start + 1)
+        Impl.CharIn(start, bitSet, NonEmptyList.one(range.start -> range.end))
+      case _ =>
+        val ary = cs.toArray
+        Arrays.sort(ary)
+        rangesFor(ary) match {
+          case NonEmptyList((low, high), Nil) if low == Char.MinValue && high == Char.MaxValue =>
```

Comment:
actually, we aren't doing this optimization above are we? if `range.start == 0` and `range.end == 0xffff` we should return `anyChar` right?